manager/src/main/java/com/redhat/service/bridge/manager/workers/resources/AbstractWorker.java

```diff
@@ -0,0 +1,181 @@
+    protected T recordAttempt(T managedResource) {
+        T resource = getDao().findById(managedResource.getId());
+        resource.getDependencyStatus().recordAttempt();
+        getDao().persist(resource);
+        return resource;
+    }
+
+    @Transactional
+    protected T setDependencyReady(T managedResource, boolean ready) {
+        T resource = getDao().findById(managedResource.getId());
+        resource.getDependencyStatus().setReady(ready);
+        getDao().persist(resource);
+        return resource;
+    }
+
+    @Transactional
+    protected T setDependencyDeleted(T managedResource, boolean deleted) {
+        T resource = getDao().findById(managedResource.getId());
+        resource.getDependencyStatus().setDeleted(deleted);
+        getDao().persist(resource);
+        return resource;
+    }
+
+    protected abstract PanacheRepositoryBase<T, String> getDao();
+
+    @Override
+    public T createDependencies(T managedResource) {
+        // Don't process resources that have their Work completed
+        if (managedResource.getStatus() == ManagedResourceStatus.READY || managedResource.getStatus() == ManagedResourceStatus.FAILED) {
+            return managedResource;
+        }
+
+        // Fail when we've had enough
+        if (areRetriesExceeded(managedResource)) {
+            info(LOGGER,
+                    String.format("Max retry attempts exceeded trying to create dependencies for '%s' [%s].",
+                            managedResource.getName(),
+                            managedResource.getId()));
+            return setStatus(managedResource, ManagedResourceStatus.FAILED);
+        }
+        if (isTimeoutExceeded(managedResource)) {
+            info(LOGGER,
+                    String.format("Timeout exceeded trying to create dependencies for '%s' [%s].",
+                            managedResource.getName(),
+                            managedResource.getId()));
+            return setStatus(managedResource, ManagedResourceStatus.FAILED);
+        }
+
+        try {
+            return runCreateOfDependencies(managedResource);
+        } catch (Exception e) {
+            info(LOGGER,
+                    String.format("Failed to create dependencies for '%s' [%s].%nDependency status: %s%n%s",
+                            managedResource.getName(),
+                            managedResource.getId(),
+                            managedResource.getDependencyStatus(),
+                            e.getMessage()));
+            // Something has gone wrong. We need to retry.
+            return recordAttempt(managedResource);
+        }
+    }
+
+    @Transactional
+    protected boolean areRetriesExceeded(T managedEntity) {
+        //This needs to be in a transaction as the child is loaded lazily
+        return getDao().findById(managedEntity.getId()).getDependencyStatus().getAttempts() > maxRetries;
+    }
+
+    @Transactional
+    protected boolean isTimeoutExceeded(T managedEntity) {
```

Comment:
Same comment here about timeout. Do we record the `scheduled` and `lastUpdated` timestamp on the `Work` item. The `Worker` for each resource type can then provide a timeout period that the work must be completed. In psuedocode, something like:

```
Work.getLastUpdated() - Work.getScheduledAt() > Worker.getTimeoutForResource()
```

Again, that encapsulates all the retry/timeout handling logic into the `Work` item.

Reply:
Move to `Work` however I still have a single global timeout.

I'll create a new JIRA for this on Monday.